pgmax/fg/graph.py

```diff
@@ -337,20 +240,28 @@ def factors(self) -> OrderedDict[Type, Tuple[nodes.Factor, ...]]:
                     tuple(
                         [
                             factor
-                            for factor_group in self.factor_groups[factor_type]
+                            for factor_group in self._factor_types_to_groups[
+                                factor_type
+                            ]
                             for factor in factor_group.factors
                         ]
                     ),
                 )
-                for factor_type in self.factor_groups
+                for factor_type in self._factor_types_to_groups
             ]
         )
         return factors
 
     @property
-    def factor_groups(self) -> OrderedDict[Type, List[groups.FactorGroup]]:
+    def factor_groups(self) -> Tuple[groups.FactorGroup, ...]:
         """Tuple of factor groups in the factor graph"""
-        return self._factor_types_to_groups
+        return tuple(
+            [
+                factor_group
+                for factor_groups_per_type in self._factor_types_to_groups.values()
+                for factor_group in factor_groups_per_type
+            ]
+        )
```

Comment:
Why this change? To be used in the hash? Looks equivalent to the previous implementation (i.e. it doesn't seem like we achieved anything with this change and the change in the hash). This also looks to be the reason you have to access a private attribute in one of the notebooks